src/lmic/radio.c

```diff
@@ -791,6 +828,85 @@ u1_t radio_rssi () {
     return r;
 }
 
+// monitor rssi for specified number of ostime_t ticks, and return statistics
+// This puts the radio into RX continuous mode, waits long enough for the
+// oscillators to start and the PLL to lock, and then measures for the specified
+// period of time.  The radio is then returned to idle.
+//
+// RSSI returned is expressed in units of dB, and is offset according to the
+// current radio setting per section 5.5.5 of Semtech 1276 datasheet.
+void radio_monitor_rssi(ostime_t nTicks, oslmic_radio_rssi_t *pRssi) {
+    uint8_t rssiMax, rssiMin;
+    uint16_t rssiSum;
+    uint16_t rssiN;
+
+    int rssiAdjust;
+    ostime_t tBegin;
+    int notDone;
+
+    rxlora(RXMODE_SCAN);
+
+    // while we're waiting for the PLLs to spin up, determine which
+    // band we're in and choose the base RSSI.
+    if (LMIC.freq > SX127X_FREQ_LF_MAX) {
+            rssiAdjust = SX127X_RSSI_ADJUST_HF;
+    } else {
+            rssiAdjust = SX127X_RSSI_ADJUST_LF;
+    }
+    rssiAdjust += hal_getRssiCal();
+
+    // zero the results
+    rssiMax = 255;
+    rssiMin = 0;
+    rssiSum = 0;
+    rssiN = 0;
+
+    // wait for PLLs
+    hal_waitUntil(os_getTime() + SX127X_RX_POWER_UP);
+
+    // scan for the desired time.
+    tBegin = os_getTime();
+    rssiMax = 0;
+
+    /* XXX
+     * In this loop, micros() in os_getTime() returns a past time sometimes.
+     * At least, it happens on Dragino LoRa Mini.
+     * the return value of micros() looks not to be stable in IRQ disabled.
+     * Once it happens, this loop never exit infinitely.
+     * In order to prevent it, it enables IRQ before calling os_getTime(),
+     * disable IRQ again after that.  in addition to, a sentry checks the
+     * number of loop for sure.
+     */
```

Comment:
Yes, we saw this too. The solution we adopted was: fix the board support package!  `micros()` should work! But I understand that this is more ambitious than most people want. I would prefer to do this in `os_gettime()`, based on a config flag. 

BTW: my guess is that the `sentry` code isn't actually working; C language definition of `int sentry = 0` at line 881 is that every pass through the block (every pass through the do loop), a new instance of `sentry` is initialized. So if we move the interrupt enable to `os_gettime()`, we can leave this code unchanged (and also fix other areas). This turns out to be a widespread problem and causes timing inaccuracies which lead to missed RX events. And the basic time-keeping problem also causes the OneWire library not to work (fixed that, too).

Reply:
The sentry was added because I didn't have confidence in my modification.  It looks not to need any more.  My code has been working for more than two weeks (about 1 minute cycle)

Reply:
OK, before making an official release, I'll like to refactor a little. I'll let you know when ready to test, later this week.